packages/syft/src/syft/service/queue/zmq_queue.py

```diff
@@ -261,12 +261,17 @@ def preprocess_action_arg(self, arg: Any) -> None:
         if res.is_err():
             return arg
         action_object = res.ok()
-        data = action_object.syft_action_data
-        new_data = self.unwrap_nested_actionobjects(data)
-        new_action_object = ActionObject.from_obj(new_data, id=action_object.id)
-        res = self.action_service.set(
-            context=self.auth_context, action_object=new_action_object
-        )
+
+        # TODO: do this elsewhere, because here it breaks mongo by setting data back
```

Comment:
+1 Ideally, we shoudn't have nested action objects, but even in the case if were to unwrap nested action object, and save it again to the Database, we should do the following:

```python

data = action_object.syft_action_data
new_data = self.unwrap_nested_actionobjects(data)
new_action_object = ActionObject.from_obj(new_data, id=action_object.id)

# this will save the actual data to blob storage and only save the metadata in mongo store.
new_action_object._save_to_blob_storage()
res = self.action_service.set(
            context=self.auth_context, action_object=new_action_object
        )




```

Reply:
cc: @koenvanderveen

Reply:
the idea was that you can do something like this

```
@syft_function
def map():
   ...
@syft_function
def reduce():
   ...

@syft_function
def outer():
   map_results=[]
   for x in [1,2,3]:
       map_job = domain.launch_job(map, x)
       map_results.append(map.job.results) # this does not exist yet, but we do have a pointer
   
   reduce_job = domain.launch_job(reduce, map_results) # this is a list of ActionObjects that do not exist yet
   return reduce_job.result
```

Reply:
agreed on how it should be stored @shubham3121. I think this was (partially?) created before we had blob storage so it was probably not changed when we introduced blob storage